File: streaming_yorkie/src/main/java/com/lethalmaus/streaming_yorkie/dao/F4FDAO.java
package com.lethalmaus.streaming_yorkie.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.lethalmaus.streaming_yorkie.entity.F4FEntity;
import com.lethalmaus.streaming_yorkie.entity.UserEntity;

/**
 * F4FDAO Interface
 * @author LethalMaus
 */
@Dao
public interface F4FDAO {

    /**
     * Get Follow4Follow UserEntity by position
     * @author LethalMaus
     * @param offset int
     * @return UserEntity
     */
    @Query("SELECT * FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    UserEntity getFollow4FollowUserByPosition(int offset);

    /**
     * Get Follow4Follow UserEntity count
     * @author LethalMaus
     * @return count int
     */
    @Query("SELECT COUNT(id) FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    int getFollow4FollowUserCount();

    /**
     * Get Followed_NotFollowing UserEntity by position
     * @author LethalMaus
     * @param offset int
     * @return UserEntity
     */
    @Query("SELECT * FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    UserEntity getFollowedNotFollowingUserByPosition(int offset);

    /**
     * Get Followed_NotFollowing UserEntity for AutoFollow
     * @author LethalMaus
     * @return UserEntity
     */
    @Query("SELECT * FROM follower WHERE (status LIKE 'NEW' OR status LIKE 'CURRENT') AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1")
    UserEntity getFollowedNotFollowingUserForAutoFollow();

    /**
     * Get Followed_NotFollowing UserEntity count
     * @author LethalMaus
     * @return count int
     */
    @Query("SELECT COUNT(id) FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    int getFollowedNotFollowingUserCount();

    /**
     * Get NotFollowed_Following UserEntity by position
     * @author LethalMaus
     * @param offset int
     * @return UserEntity
     */
    @Query("SELECT * FROM following WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    UserEntity getNotFollowedFollowingUserByPosition(int offset);

    /**
     * Get NotFollowed_Following UserEntity for AutoFollow
     * @author LethalMaus
     * @return UserEntity
     */
    @Query("SELECT * FROM following WHERE (status LIKE 'NEW' OR status LIKE 'CURRENT') AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1")
    UserEntity getNotFollowedFollowingUserForAutoFollow();

    /**
     * Get NotFollowed_Following UserEntity count
     * @author LethalMaus
     * @return count int
     */
    @Query("SELECT COUNT(id) FROM following WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    int getNotFollowedFollowingUserCount();

    /**
     * Get excluded Follow4Follow UserEntity by position
     * @author LethalMaus
     * @param offset int
     * @return UserEntity
     */
    @Query("SELECT * FROM f4f_excluded ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    F4FEntity getExcludedFollow4FollowUserByPosition(int offset);

    /**
     * Get excluded Follow4Follow UserEntity count
     * @author LethalMaus
     * @return count int
     */
    @Query("SELECT COUNT(id) FROM f4f_excluded")
    int getExcludedFollow4FollowUserCount();

    /**
     * Delete UserEntity by Id
     * @author LethalMaus
     * @param id int
     */
    @Query("DELETE FROM f4f_excluded WHERE id = :id")
    void deleteUserById(int id);

    /**
     * Inserts a UserEntity and replaces on conflict
     * @author LethalMaus
     * @param f4FEntity UserEntity
     */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertUser(F4FEntity f4FEntity);
}

File: app/src/main/java/com/lethalmaus/streaming_yorkie/database/dao/F4FDAO.kt
package com.lethalmaus.streaming_yorkie.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.lethalmaus.streaming_yorkie.database.entity.F4FEntity
import com.lethalmaus.streaming_yorkie.database.entity.FollowerEntity
import com.lethalmaus.streaming_yorkie.database.entity.FollowingEntity

@Dao
interface F4FDAO {

    @Query("SELECT * FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    fun getFollow4FollowUserByPosition(offset: Int): FollowerEntity?

    @get:Query("SELECT COUNT(id) FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    val follow4FollowUserCount: Int

    @Query("SELECT * FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    fun getFollowedNotFollowingUserByPosition(offset: Int): FollowerEntity?

    @get:Query("SELECT * FROM follower WHERE (status LIKE 'NEW' OR status LIKE 'CURRENT') AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1")
    val followedNotFollowingUserForAutoFollow: FollowerEntity?

    @get:Query("SELECT COUNT(id) FROM follower WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM following WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    val followedNotFollowingUserCount: Int

    @Query("SELECT * FROM following WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    fun getNotFollowedFollowingUserByPosition(offset: Int): FollowingEntity?

    @get:Query("SELECT * FROM following WHERE (status LIKE 'NEW' OR status LIKE 'CURRENT') AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded) ORDER BY created_at DESC LIMIT 1")
    val notFollowedFollowingUserForAutoFollow: FollowingEntity?

    @get:Query("SELECT COUNT(id) FROM following WHERE status NOT LIKE 'UNFOLLOWED' AND id NOT IN (SELECT id FROM follower WHERE status NOT LIKE 'UNFOLLOWED') AND id NOT IN (SELECT id FROM f4f_excluded)")
    val notFollowedFollowingUserCount: Int

    @Query("SELECT * FROM f4f_excluded ORDER BY created_at DESC LIMIT 1 OFFSET :offset")
    fun getExcludedFollow4FollowUserByPosition(offset: Int): F4FEntity?

    @get:Query("SELECT COUNT(id) FROM f4f_excluded")
    val excludedFollow4FollowUserCount: Int

    @Query("DELETE FROM f4f_excluded WHERE id = :id")
    fun deleteUserById(id: Int)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertUser(f4FEntity: F4FEntity?)
}